staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate.go

```diff
@@ -273,19 +300,29 @@ func NegotiateMediaTypeOptions(header string, accepted []AcceptedMediaType, endp
 	}
 
 	clauses := goautoneg.ParseAccept(header)
+	candidateMediaTypeList := []candidateMediaType{}
 	for _, clause := range clauses {
 		for i := range accepted {
 			accepts := &accepted[i]
 			switch {
 			case clause.Type == accepts.Type && clause.SubType == accepts.SubType,
 				clause.Type == accepts.Type && clause.SubType == "*",
 				clause.Type == "*" && clause.SubType == "*":
-				// TODO: should we prefer the first type with no unrecognized options?  Do we need to ignore unrecognized
-				// parameters.
-				return acceptMediaTypeOptions(clause.Params, accepts, endpoint)
+				candidateMediaTypeList = append(candidateMediaTypeList, candidateMediaType{accepted: accepts, clauses: clause})
 			}
 		}
 	}
+
+	//run until element all removed
+	for len(candidateMediaTypeList) > 0 {
+		val, index := mostSpecificMediaType(candidateMediaTypeList)
+		if retVal, ret := acceptMediaTypeOptions(val.clauses.Params, val.accepted, endpoint); ret == true {
+			return retVal, true
+		}
+		//remove element
+		candidateMediaTypeList = append(candidateMediaTypeList[:index], candidateMediaTypeList[index+1:]...)
```

Comment:
It would be better to swap index with the last item, then truncate the array by one.  This code will be called on every request and some efficiency matters.